### PR TITLE
DENG-651 Aggregates with 2023 KPIs

### PR DESCRIPTION
**Pull request by lucia-vargas-a**
[DENG-651](https://mozilla-hub.atlassian.net/browse/DENG-651). This PR implements the 2023 KPI QDAU based on this [data model](https://docs.google.com/document/d/14dJyFAocxovHzSSSJY4mSM3aEadkdPqAtV9X4olCEGA/edit?usp=sharing) (which is part of a POC for the Data Modeling process proposal).

[DENG-654](https://mozilla-hub.atlassian.net/browse/DENG-654) This PR also updates active_users_aggregates_attribution_v1 to use the new table fenix.firefox_android_clients to retrieve Fenix client's attribution. This is important before running the backfill.

[Data checks](https://docs.google.com/spreadsheets/d/1FvrLS0d5ppDqVvjdFkC2US4L2QtZwsfE2kDFmL2XWEI/edit?usp=share_link) in BigQuery analysis tables to validate that current metrics remain correct, tested only 2023-01-01.

Before merging this PR, it is required to deploy the schemas to add the new column.

- [x] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [x] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [x] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

[DENG-651]: https://mozilla-hub.atlassian.net/browse/DENG-651?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DENG-654]: https://mozilla-hub.atlassian.net/browse/DENG-654?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/query.sql [hunk 1]
```diff
@@ -69,6 +69,16 @@ SELECT
   COUNT(DISTINCT IF(days_since_seen = 0, client_id, NULL)) AS dau,
   COUNT(DISTINCT IF(days_since_seen < 7, client_id, NULL)) AS wau,
   COUNT(DISTINCT client_id) AS mau,
+  COUNT(
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/query.sql:72]: Using a DISTINCT here for unique client count, which slightly differs from the [proposed query](https://sql.telemetry.mozilla.org/queries/89768/source).

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql [hunk 2]
```diff
@@ -50,130 +27,13 @@ WITH baseline AS (
...
-    base.search_count,
-    base.search_with_ads,
-    base.uri_count,
-    base.active_hours_sum
-  FROM
-    baseline AS base
+    `moz-fx-data-shared-prod.telemetry.unified_metrics` AS unified_metrics
   LEFT JOIN
-    fenix_unique_attribution AS att
+    fenix.firefox_android_clients AS att
```
Comment (lucia-vargas-a) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql:32]: Raplacement to use `fenix.firefox_android_clients` to retrieve first unique attribution.

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/schema.yaml [hunk 3]
```diff
@@ -76,3 +76,6 @@ fields:
 - mode: NULLABLE
   name: attributed
   type: BOOLEAN
+- mode: NULLABLE
```
Comment (wwyc) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/schema.yaml:79]: should a description be added?

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/schema.yaml [hunk 4]
```diff
@@ -74,3 +74,6 @@ fields:
 - mode: NULLABLE
   name: active_hours
   type: FLOAT
+- mode: NULLABLE
```
Comment (wwyc) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/schema.yaml:77]: see comment above re description

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/schema.yaml [hunk 5]
```diff
@@ -83,3 +83,6 @@ fields:
 - mode: NULLABLE
   name: language_name
   type: STRING
+- mode: NULLABLE
```
Comment (wwyc) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/schema.yaml:86]: see comment above re description

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/metadata.yaml [hunk 6]
```diff
@@ -35,3 +35,4 @@ bigquery:
     - city
     - app_name
     - attribution_medium
+references: {}
```
Comment (wwyc) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/metadata.yaml:38]: Is this needed?

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/metadata.yaml [hunk 7]
```diff
@@ -27,3 +27,4 @@ bigquery:
     - app_name
     - attribution_medium
     - channel
+references: {}
```
Comment (wwyc) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/metadata.yaml:30]: Is this needed?

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/schema.yaml (hunk 3 above)
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/schema.yaml:79]: Definitely!
I've added the description from the Use Story. WE can update it with the link to the 'source of truth' once we have it.

**Comment by irrationalagent** [comment]
Kind of a bigger picture comment here (sorry). 

We are beginning to work on our KR related to improving the source of truth and lineage of our KPIs. As part of that, we are starting to think through possible alternative aggregation models that would reduce the number of steps in the ETL pipeline needed to report accurate KPIs. Just as an example (by NO means a proposal at this point), we have talked about a model where we aggregate into per-app tables first directly from clients_last_seen (rather from unified metrics, which we are doing now), then combine if needed into a unioned table of aggregates from all the apps/platforms.

On the other hand, the work's been done so maybe there is a case to do this now as a stopgap that would allow more efficient qdau queries for the time being.

Talking with @bochocki , we are going to setup a meeting with you and George about this to work things out. So let's hold off on a decision about this until then.

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql (hunk 2 above)
Comment (wwyc) [reply sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql:32]: It looks like the `fenix.firefox_android_client` table can have more than one row per `client_id`.  Therefore, this left join can result in more than one row per `client_id`.  Is this be the expected behaviour?

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/query.sql [hunk 8]
```diff
@@ -19,6 +19,16 @@ SELECT
...
+  COUNT(
+    DISTINCT IF(
+      days_since_seen = 0
+      AND normalized_app_name = 'Firefox Desktop'
+      AND active_hours_sum > 0
+      AND uri_count > 0,
+      client_id,
+      NULL
+    )
+  ) AS qdau,
```
Comment (bochocki) [inline sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/query.sql:31]: Nothing to do now, but it feels a bit fragile to me that this logic is defined in three different places:
- sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql
- sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/query.sql
- sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/query.sql

I'm still learning my way around this repo, but is there any chance that future work could consolidate this code in a single place, possibly upstream?

**Review by bochocki — APPROVED** [review]
This looks good to me! I left some comments, but they're more for my own curiosity than anything.

Following up on @irrationalagent's comment, we met today (@lucia-vargas-a, @wwyc, @gkabbz, @irrationalagent) and it sounds like there's possibly some upcoming work to re-think how we implement aggregations in the pipeline. That said, this seems like a short-term win in terms of making it easier to get `qdau` out of the existing pipeline.

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/schema.yaml [hunk 9]
```diff
@@ -83,3 +83,7 @@ fields:
 - mode: NULLABLE
   name: language_name
   type: STRING
+- description: Count of qualified clients sending a main ping on a given day
```
[the remark this answers is not in the document]
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/schema.yaml:86]: That's how it works, and it's much more specific. Thank you!

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/query.sql (hunk 8 above)
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/query.sql:31]: That's the plan for a future version yes, for all the metrics ideally e.g. searches, ad clicks, active users. Using [metric hub](https://docs.google.com/presentation/d/1sXyekWMhYfXAwvEsCT5Q4Cjnu5owQ8LJ-VGyX7gGwCk/edit#slide=id.g82761e80df_0_1948) is the current proposal for this consolidation.

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql (hunk 2 above)
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_attribution_v1/query.sql:32]: Looking into this issue, we need to solve it from the upstream `moz-fx-data-shared-prod_mozilla_firefox_derived.baseline_clients_first_seen`, which potentially requires a backfill itself before we backfill `active_users_aggregates_attribution`. We don't require changes here, but in the related [Jira Task](https://mozilla-hub.atlassian.net/jira/software/c/projects/DENG/boards/465?modal=detail&selectedIssue=DENG-656).

File: sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/schema.yaml [hunk 10]
```diff
@@ -74,3 +74,7 @@ fields:
 - mode: NULLABLE
   name: active_hours
   type: FLOAT
+- description: Count of qualified clients sending a main ping on a given day
```
[the remark this answers is not in the document]
Comment (lucia-vargas-a) [reply sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_device_v1/schema.yaml]: LGTM! I added to the description that it is only counting desktop users (we can change it if we start counting mobile or create a different column) and specified that mobile DAU = QDAU. Also clarified in the column name  for clarity.

**Pull request merged**